src/analysis/analysis-service.js

```diff
@@ -19,21 +19,48 @@ var AnalysisService = function (opts) {
   this._apiKey = opts.apiKey;
   this._authToken = opts.authToken;
   this._camshaftReference = opts.camshaftReference || camshaftReference; // For testing purposes
-  this._layersCollection = opts.layersCollection;
-  this._dataviewsCollection = opts.dataviewsCollection;
+  this._layersCollection = opts.layersCollection || [];
+  this._dataviewsCollection = opts.dataviewsCollection || [];
 };
 
 /**
- * Create an analysis node
- *
- * Recursively generates a graph of analyses and returns the "origin" node. Each node
- * may have one or more "source" params pointing to another node. If a node had been created
- * already, this method updates the attributes of the existing node. New nodes are added to
- * the collection of analyses that has been injected.
+  * Recursively generates a graph of analyses and returns the "origin" node. Each node
+  * may have one or more "source" params pointing to another node. If a node had been created
+  * already, this method updates the attributes of the existing node. New nodes are added to
+  * the collection of analyses that has been injected.
+  */
+AnalysisService.prototype.analyse = function (analysisDefinition) {
```

Comment:
Could we use `createAnalysis` and `updateAnalysis` here too?

```
AnalysisService.prototype.analyse = function (analysisDefinition) {
  var analysis = this.findNodeById(analysisDefinition.id);
  if (analysis) {
    this.updateAnalysis(analysisDefinition);
  } else {
    analysis = this.createAnalysis(analysisDefinition);
  }
  return analysis;
}
```

Reply:
I don't think so, because the analyse method must be applied recursively to the sourceNodes. Your approach executes the analyse method only in the first iteration

Reply:
`this.createAnalysis` creates analysis nodes recursively... Doesn't it?

Reply:
Yes but it applies recursive `createAnalysis`,  no recursisve `analyse`